clang/test/SemaCXX/ext-int.cpp

```diff
@@ -0,0 +1,228 @@
+  _ExtInt(3.0) x; // expected-error {{integral constant expression must have integral or unscoped enumeration type, not 'double'}}
+
+  return 0;
+}
+
+template <_ExtInt(5) I> // TODO: CodeGen test for mangling.
+struct ExtIntTemplParam {
+  static constexpr _ExtInt(5) Var = I;
+};
+
+template<typename T>
+void deduced_whole_type(T){}
+template<int I>
+void deduced_bound(_ExtInt(I)){}
+
+// Ensure ext-int can be used in template places.
+void Templates() {
+  ExtIntTemplParam<13> a;
+  constexpr _ExtInt(3) b = 1;
+  ExtIntTemplParam<b> c;
+  constexpr _ExtInt(9) d = 1;
+  ExtIntTemplParam<b> e;
+
+  deduced_whole_type(b);
+  deduced_bound(b);
+}
+
+template <typename T, typename U>
+struct is_same {
+  static constexpr bool value = false;
+};
+template <typename T>
+struct is_same<T,T> {
+  static constexpr bool value = true;
+};
+
+// Test result types of Unary/Bitwise/Binary Operations:
+void Ops() {
+  _ExtInt(43) x43_s = 1, y43_s = 1;
+  _ExtInt(sizeof(int) * 8) x32_s = 1, y32_s = 1;
+  unsigned _ExtInt(sizeof(unsigned) * 8) x32_u = 1, y32_u = 1;
+  _ExtInt(4) x4_s = 1, y4_s = 1;
+  unsigned _ExtInt(43) x43_u = 1, y43_u = 1;
+  unsigned _ExtInt(4) x4_u = 1, y4_u = 1;
+  int x_int = 1, y_int = 1;
+  unsigned x_uint = 1, y_uint = 1;
+  bool b;
+
+  // Same size/sign ops don't change type.
+  static_assert(is_same<decltype(x43_s + y43_s), _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x4_s - y4_s), _ExtInt(4)>::value,"");
+  static_assert(is_same<decltype(x43_u * y43_u), unsigned _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x4_u / y4_u), unsigned _ExtInt(4)>::value,"");
+
+  // Same size/diff sign goes to unsigned.
+  static_assert(is_same<decltype(x43_u + y43_s), unsigned _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x4_s - y4_u), unsigned _ExtInt(4)>::value,"");
+  static_assert(is_same<decltype(x43_s * y43_u), unsigned _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x4_u / y4_s), unsigned _ExtInt(4)>::value,"");
+
+  // Different size prefers largest.
+  static_assert(is_same<decltype(x43_s + y4_s), _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x43_s - y4_u), _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x43_u * y4_u), unsigned _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x4_u / y43_u), unsigned _ExtInt(43)>::value,"");
+
+  // When mixed with standard types, largest prefered, then standard types 
+  // > ext_int.
+  static_assert(is_same<decltype(x43_s + x_int), _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x43_u - x_int), unsigned _ExtInt(43)>::value,"");
+  static_assert(is_same<decltype(x32_s * x_int), int>::value,"");
+  static_assert(is_same<decltype(x32_u / x_int), unsigned int>::value,"");
+  static_assert(is_same<decltype(x32_s * x_uint), unsigned int>::value,"");
+  static_assert(is_same<decltype(x32_u / x_uint), unsigned int>::value,"");
```

Comment:
Maybe I missed it but you probably want a case with a char mixed with an ExtInt of size < 32 because it's a weird special case where we don't promote all the way to int.